cluster/src/main/java/io/atomix/cluster/AtomixCluster.java

```diff
@@ -215,6 +218,7 @@ public ClusterEventService getEventService() {
     return messagingService.start()
         .thenComposeAsync(v -> broadcastService.start(), threadContext)
         .thenComposeAsync(v -> membershipService.start(), threadContext)
+        .thenComposeAsync(v -> locationProvider.join(membershipService.getLocalMember().address()), threadContext)
```

Comment:
If I understand correctly, in case that one of the composed calls fail, we'll be in a corrupted state where some services are started and some others are stopped.

Shouldn't we try these operations to be atomic?

Reply:
that's exactly what I mean, if `messagingService.start()` succeeds but `broadcastService.start()` fails, `messagingService` will be started while it would be better to shut it down.